### PR TITLE
docs(headless commerce react samples): Add instant products

### PR DESCRIPTION
**Pull request by fbeaudoincoveo**
https://coveord.atlassian.net/browse/KIT-3431

### CODE REVIEW

File: packages/samples/headless-commerce-react/src/components/instant-products/instant-products.tsx [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  controller: HeadlessInstantProducts;
+  navigate: (pathName: string) => void;
+}
+
+export default function InstantProducts(props: IInstantProductProps) {
+  const {controller, navigate} = props;
+  const [state, setState] = useState(controller.state);
+
+  useEffect(
+    () => controller.subscribe(() => setState({...controller.state})),
```
Comment (Spuffynism) [inline packages/samples/headless-commerce-react/src/components/instant-products/instant-products.tsx:17]: Why do we spread the state?
```suggestion
    () => controller.subscribe(() => setState(controller.state)),
```

File: packages/samples/headless-commerce-react/src/components/search-box/search-box.tsx [hunk 2]
```diff
@@ -15,31 +21,65 @@ export default function SearchBox(props: ISearchBoxProps) {
...
+    instantProductsController.updateQuery(e.target.value);
+  };
+
   return (
     <div className="Searchbox">
       <input
         className="SearchBoxInput"
         value={state.value}
-        onChange={(e) => controller.updateText(e.target.value)}
+        onChange={(e) => onSearchBoxInputChange(e)}
```
Comment (Spuffynism) [inline packages/samples/headless-commerce-react/src/components/search-box/search-box.tsx:34]:
```suggestion
        onChange={onSearchBoxInputChange}
```

File: packages/samples/headless-commerce-react/src/components/standalone-search-box/standalone-search-box.tsx [hunk 3]
```diff
@@ -25,12 +30,17 @@ export default function StandaloneSearchBox(props: IStandaloneSearchBoxProps) {
...
+    instantProductsController.updateQuery(e.target.value);
+  };
+
   return (
     <div className="StandaloneSearchbox">
       <input
         className="SearchBoxInput"
         value={state.value}
-        onChange={(e) => controller.updateText(e.target.value)}
+        onChange={(e) => onSearchBoxInputChange(e)}
```
Comment (Spuffynism) [inline packages/samples/headless-commerce-react/src/components/standalone-search-box/standalone-search-box.tsx:43]:
```suggestion
        onChange={onSearchBoxInputChange}
```

**Review by Spuffynism — APPROVED** [review]

File: packages/samples/headless-commerce-react/src/components/search-box/search-box.tsx [hunk 4]
```diff
@@ -15,31 +21,65 @@ export default function SearchBox(props: ISearchBoxProps) {
...
+            <th>Instant products</th>
+          </tr>
+        </thead>
+        <tbody>
+          <tr>
+            <td>
+              {state.suggestions.length > 0 && (
+                <ul className="QuerySuggestions">
+                  {state.suggestions.map((suggestion, index) => (
+                    <li key={index} className="QuerySuggestion">
```
Comment (Spuffynism) [inline packages/samples/headless-commerce-react/src/components/search-box/search-box.tsx:55]: Should we use the `rawValue` as key actually? It should be unique. An `index` could be the same for two different suggestions, on different queries submitted to the search box.
```suggestion
                    <li key={rawValue} className="QuerySuggestion">
```

Comment (louis-bompart) [reply packages/samples/headless-commerce-react/src/components/search-box/search-box.tsx:55]: Cannot be the same @Spuffynism , see [📖](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map#callbackfn) `Array.prototype.map`

File: packages/samples/headless-commerce-react/src/components/standalone-search-box/standalone-search-box.tsx (hunk 3 above)
Comment (louis-bompart) [reply packages/samples/headless-commerce-react/src/components/standalone-search-box/standalone-search-box.tsx:43]: I think this works here, but be careful with [`this`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/this)

File: packages/samples/headless-commerce-react/src/components/search-box/search-box.tsx (hunk 4 above)
Comment (Spuffynism) [reply packages/samples/headless-commerce-react/src/components/search-box/search-box.tsx:55]: What I mean is that two different arrays share indices for items across renders. Using the `rawValue` would guarantee unique items, for unique values.

```
// first render, for query "cat"
// 0       1              2
  ["cats", "caterpillar", "category"]

// second render, for query "cate"
// 0              1
  ["caterpillar", "category"]
```

> A well-chosen key provides more information than the position within the array. Even if the position changes due to reordering, the key lets React identify the item throughout its lifetime.

https://react.dev/learn/rendering-lists#why-does-react-need-keys

File: packages/samples/headless-commerce-react/src/components/instant-products/instant-products.tsx (hunk 1 above)
Comment (fbeaudoincoveo) [reply packages/samples/headless-commerce-react/src/components/instant-products/instant-products.tsx:17]: Oh yeah I just started from the result list component as a boilerplate, and for some reason I spread the state there...

No need to spread it :D

File: packages/samples/headless-commerce-react/src/components/search-box/search-box.tsx (hunk 4 above)
Comment (fbeaudoincoveo) [reply packages/samples/headless-commerce-react/src/components/search-box/search-box.tsx:55]: Yep, I'll use better keys

File: packages/samples/headless-commerce-react/src/components/instant-products/instant-products.tsx (hunk 1 above)
Comment (alexprudhomme) [reply packages/samples/headless-commerce-react/src/components/instant-products/instant-products.tsx:17]: For state objects its better to spread to make sure the nested properties changed are detected. [here](https://react.dev/reference/react/useState#examples-objects)

**Review by alexprudhomme — APPROVED** [review]

File: packages/samples/headless-commerce-react/src/components/instant-products/instant-products.tsx (hunk 1 above)
Comment (fbeaudoincoveo) [reply packages/samples/headless-commerce-react/src/components/instant-products/instant-products.tsx:17]: OK so I apparently accidentally applied a best practice then hehe

File: packages/samples/headless-commerce-react/src/components/instant-products/instant-products.tsx [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  return (
+    <div className="InstantProducts">
+      {state.products.length === 0 && (
+        <span>
+          No instant products for query <b>{state.query}</b>
+        </span>
+      )}
+      <p>
+        Instant products for query <b>{state.query}</b>
+      </p>
```
Comment (Spuffynism) [inline packages/samples/headless-commerce-react/src/components/instant-products/instant-products.tsx:49]: Should we put this all in a ternary? To not have both the "No instant products" and the "Instant products" messages shown at the same time?

**Review by Spuffynism — APPROVED** [review]

**Review by fpbrault — APPROVED** [review]
LGTM

File: packages/samples/headless-commerce-react/src/components/search-box/search-box.tsx [hunk 6]
```diff
@@ -15,31 +21,65 @@ export default function SearchBox(props: ISearchBoxProps) {
...
+        <thead>
+          <tr>
+            <th>Query suggestions</th>
+            <th>Instant products</th>
+          </tr>
+        </thead>
+        <tbody>
+          <tr>
+            <td>
+              {state.suggestions.length > 0 && (
```
[the remark this answers is not in the document]
Comment (fbeaudoincoveo) [reply packages/samples/headless-commerce-react/src/components/search-box/search-box.tsx]: I made significant changes to the way this works, so this comment is no longer relevant.

**Pull request merged**